packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -440,7 +451,9 @@ class _HelperErrorState extends State<_HelperError> with SingleTickerProviderSta
             begin: const Offset(0.0, -0.25),
             end: Offset.zero,
           ).evaluate(_controller.view),
-          child: widget.error ?? Text(
+          child: widget.error
+              ?? widget.errorBuilder?.call(widget.errorText!)
```

Comment:
This implies that if errorBuilder is non-null, then errorText must also be non-null. Can you add an assertion for that? And we should probably mention it in the docs for errorBuilder too.

Reply:
There is already an assert at the top of this function to confirm that either error or errorText is provided.

Reply:
What I mean is slightly different I think: 

`assert(errorBuilder == null || errorText != null, 'errorBuilder must be used with errorText')`

And I think that should go in InputDecoration's initializers.

Reply:
Adding it at constructors won't work. errorBuilder will always be provided, and whether or not errorText exists depends on the validator (which is kind of the point of this PR). What we need to guard against is that errorBuilder should not be called without an errorText.

Reply:
Ah you're right, errorText can be null, and we can't distinguish between null and not passed at all here.

Reply:
Good discussion here 👍

I will say: even if they can't be moved to a public class, moving some of these asserts to the `_HelperError` constructor would make debugging a bit easier, since there would be a shorter stacktrace to follow.